src/arc.rs

```diff
@@ -183,47 +211,74 @@ impl PathArc {
+
+        fn maybe_init_res(res: &mut PathBuf, resolvee: &PathArc) -> Result<()> {
+            if res.as_os_str() != "" {
+                // res has already been initialized, let's leave it alone.
+                return Ok(());
             }
-        }}}
 
-        handle_prefix(self, &mut stack, &mut components, false)?;
+            // res has not been initialized, let's initialize it to the
+            // canonicalized current directory.
+            let cwd = env::current_dir().map_err(|e| {
+                Error::new(
+                    e,
+                    "getting current_dir while resolving absolute",
+                    resolvee.clone(),
+                )
+            })?;
+            *res = cwd.canonicalize().map_err(|e| {
+                Error::new(e, "canonicalizing", PathArc::new(&cwd))
+            })?;
+
+            Ok(())
+        };
+
+        for each in self.components() {
+            match each {
+                Component::Prefix(p) => {
+                    // We don't care what's already in res, we can entirely
+                    // replace it..
+                    res = make_verbatim_prefix(&p)?;
+                }
 
-        for component in components {
-            match component {
-                Component::CurDir => { /* ignore, probably impossible */ }
-                Component::Prefix(_) => unreachable!(),
                 Component::RootDir => {
-                    if cfg!(unix) {
-                        unreachable!("root is already handled on unix");
+                    if cfg!(windows) {
+                        // A root path component is relative to some existing
+                        // path, so we may need to initialize res.
+                        maybe_init_res(&mut res, self)?;
+                        res.push(each);
+                    } else {
+                        // On other platforms, a root path component is always
+                        // absolute so we can replace whatever's in res.
+                        res = Path::new(&each).to_path_buf();
                     }
-                    // This is actually possible on windows because root is distinct
-                    // from prefix (?)
-                    stack.push(to_os(component));
                 }
-                Component::ParentDir => pop_stack!(),
-                Component::Normal(_) => stack.push(to_os(component)),
-            }
-        }
 
-        if stack.is_empty() {
-            return Err(Error::new(
-                io::Error::new(io::ErrorKind::NotFound, "resolving resulted in empty path"),
-                "resolving absolute",
-                self.clone(),
-            ));
+                // This does nothing and can be ignored.
+                Component::CurDir => (),
+
+                Component::ParentDir => {
+                    // A parent component is always relative to some existing
+                    // path.
+                    maybe_init_res(&mut res, self)?;
```

Comment:
Okay, I'm starting to understand the idea here. For several of these branches `res` has to be already initialized or you have to "initialize `res` as `current_dir`. That makes sense.